core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/unmarshall/XmlProtocolUnmarshaller.java

```diff
@@ -79,44 +78,46 @@ public <TypeT extends SdkPojo> TypeT unmarshall(SdkPojo sdkPojo,
+                    // SDK bytes bound directly to the payload field should never be left empty
+                    field.set(sdkPojo, SdkBytes.fromByteArrayUnsafe(new byte[0]));
+                } else if (field.marshallingType() == MarshallingType.STRING && field.containsTrait(PayloadTrait.class)) {
+                    field.set(sdkPojo, "");
                 }
+                continue;
+            }
+
+            if (root != null && isAttribute(field)) {
+                root.getOptionalAttributeByName(field.unmarshallLocationName())
+                    .ifPresent(e -> field.set(sdkPojo, e));
+                continue;
+            }
+
+            List<XmlElement> element = isExplicitPayloadMember(field) ?
+                                       singletonList(root) :
+                                       root.getElementsByName(field.unmarshallLocationName());
 
-                if (isAttribute(field)) {
-                    root.getOptionalAttributeByName(field.unmarshallLocationName())
-                        .ifPresent(e -> field.set(sdkPojo, e));
+            if (!CollectionUtils.isNullOrEmpty(element)) {
+                if (isExplicitPayloadMember(field) && field.marshallingType() == MarshallingType.SDK_BYTES) {
+                    field.set(sdkPojo, SdkBytes.fromInputStream(context.response().content().get()));
+                } else if (isExplicitPayloadMember(field) && field.marshallingType() == MarshallingType.STRING) {
+                    setExplicitStringPayload(unmarshaller, context, sdkPojo, element, field);
                 } else {
-                    List<XmlElement> element = isExplicitPayloadMember(field) ?
-                                               singletonList(root) :
-                                               root.getElementsByName(field.unmarshallLocationName());
-
-                    if (!CollectionUtils.isNullOrEmpty(element)) {
-                        boolean isFieldBlobTypePayload = payloadMemberAsBlobType.isPresent() &&
-                                                         payloadMemberAsBlobType.get().equals(field);
-                        if (isFieldBlobTypePayload) {
-                            field.set(sdkPojo, SdkBytes.fromInputStream(context.response().content().get()));
-                        } else {
-                            Object unmarshalled = unmarshaller.unmarshall(context, element, (SdkField<Object>) field);
-                            field.set(sdkPojo, unmarshalled);
-                        }
-                    }
+                    Object unmarshalled = unmarshaller.unmarshall(context, element, (SdkField<Object>) field);
+                    field.set(sdkPojo, unmarshalled);
                 }
-            } else {
-                Object unmarshalled = unmarshaller.unmarshall(context, null, (SdkField<Object>) field);
-                field.set(sdkPojo, unmarshalled);
             }
```

Comment:
I still can't convince myself that root = null is correct. 

It seems non-obvious that root won't be null when we call `root.getElementsByName`. Also, before this change when when `root == null`, we'd do `unmarshaller.unmarshall(context, null, (SdkField<Object>) field)`, but now we can do `unmarshaller.unmarshall(context, singletonList(null), (SdkField<Object>) field)` which feels wrong.

How about this? Not sure if it's correct since I just wrote it really quickly, but it treats `root == null` in a way that I'm less scared of.

```java
if (field.location() != MarshallLocation.PAYLOAD) {
    Object unmarshalled = unmarshaller.unmarshall(context, null, (SdkField<Object>) field);
    field.set(sdkPojo, unmarshalled);
    continue;
}

if (isExplicitPayloadMember(field)) {
    InputStream content = context.response().content().orElse(null);
    if (field.marshallingType() == MarshallingType.SDK_BYTES) {
        SdkBytes value = content == null ? SdkBytes.fromByteArrayUnsafe(new byte[0])
                                         : SdkBytes.fromInputStream(content);
        field.set(sdkPojo, value);
        continue;
    } else if (field.marshallingType() == MarshallingType.STRING) {
        // TODO: If we ever break protected APIs, just parse this as a string and remove XML-wrapping 
        // compatibility for S3.
        if (content == null) {
            field.set(sdkPojo, "");
        } else {
            setExplicitStringPayload(unmarshaller, context, sdkPojo, element, field);
        }
        continue;
    }
}

if (root == null) {
    Object unmarshalled = unmarshaller.unmarshall(context, null, (SdkField<Object>) field);
    field.set(sdkPojo, unmarshalled);
    continue;
}

if (isAttribute(field)) {
    root.getOptionalAttributeByName(field.unmarshallLocationName())
        .ifPresent(e -> field.set(sdkPojo, e));
    continue;
}

// TODO: Can this be handled in the earlier "if (isExplicitPayloadMember(field))" statements?
if (isExplicitPayloadMember(field)) {
    Object unmarshalled = unmarshaller.unmarshall(context, singletonList(root), (SdkField<Object>) field);
    field.set(sdkPojo, unmarshalled);
    continue;
}

List<XmlElement> element = root.getElementsByName(field.unmarshallLocationName());
if (!CollectionUtils.isNullOrEmpty(element)) {
    Object unmarshalled = unmarshaller.unmarshall(context, element, (SdkField<Object>) field);
    field.set(sdkPojo, unmarshalled);
}
```

Reply:
+1 updating to handle null root more explicitly per above, with a few additional tweaks